src/achievements.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+	Copyright (C) 2003 - 2022
+	by David White <dave@whitevine.net>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#include "achievements.hpp"
+
+#include "filesystem.hpp"
+#include "game_config.hpp"
+#include "log.hpp"
+#include "preferences/general.hpp"
+#include "serialization/parser.hpp"
+#include "serialization/preprocessor.hpp"
+
+static lg::log_domain log_config("config");
+#define ERR_CONFIG LOG_STREAM(err, log_config)
+
+// TODO: horizontal scrollbar is present despite `horizontal_scrollbar_mode = "never"`
+// TODO: how to make the initial height more than a single achievement? hard code the height? add more achievements?
+// TODO: make checkbox to indicate which achievements have been achieved disabled/untoggleable
+// TODO: GUIWindowDefinitionWML in achievements_dialog.hpp
+// TODO: some way to display in-game when something is achieved?
+
+// TODO: handling translatable strings? needs tstring?
+// TODO: can content_for be translatable? probably not since it's used internally for things - need two separate attributes, one translatable and one not translatable?
+
+// TODO: add to schema
```

Comment:
Although this sounds important, from your description it doesn't seem to be crucial. You essentially just need to define an entirely new schema, right? These `achievements.cfg` are not following the game config schema like most of our other WML.

Reply:
Alright, that's good. I was wondering if this was even something it could do, given the different way it loads the WML.

Reply:
Well, any WML can be validated, certainly. But if it's not included from the core _main.cfg, then it would be a separate schema. We have a number of additional schemas already, so adding one for achievements would be quite doable (it doesn't seem like it'll be a very complex schema anyway).

Reply:
This has been added.